README.md
Mesos Cookbook
==============
[![Cookbook](http://img.shields.io/cookbook/v/mesos.svg)](https://supermarket.chef.io/cookbooks/mesos)
[![Build Status](https://secure.travis-ci.org/mdsol/mesos_cookbook.png?branch=master)](http://travis-ci.org/mdsol/mesos_cookbook)
[![Gitter chat](https://img.shields.io/badge/Gitter-mdsol%2Fmesos_cookbook-brightgreen.svg)](https://gitter.im/mdsol/mesos_cookbook)

Application cookbook for installing the [Apache Mesos][] cluster manager.
Mesos provides efficient resource isolation and sharing across distributed
applications, or frameworks.  This cookbook installs Mesos via packages
provided by [Mesosphere][].

Requirements
------------

- Chef >= 12.14.60

Platform
--------
Tested on

* Ubuntu 16.04
* Ubuntu 15.10
* Ubuntu 14.04
* Ubuntu 12.04
* Debian 8.6 (with Java 7)
* CentOS 7.2
* CentOS 6.8

Supported Mesos versions
------------------------

This cookbook is tested against the following Apache Mesos versions:

* 1.3.0
* 1.1.0
* 1.0.1
* 1.0.0

We intend to support at most the three latest versions of Apache Mesos.

Attributes
----------
In order to keep the README managable and in sync with the attributes, this
cookbook documents attributes inline. The usage instructions and default
values for attributes can be found in the individual attribute files.

Configuring Mesos via attributes
-----------------------------------------
This cookbook introduces a few points of validation to prevent passing Mesos
invalid configuration options. The ruby block
`mesos-slave-configuration-validation` and
`mesos-master-configuration-validation` extract a hash of all valid Mesos
configuration options from the `--help` output of the master and slave binary
and check it against the provided attributes. This cookbook will fail to
converge if you try to use an invalid configuration option as a command line
flag attribute under `['mesos']['master']['flags']`
or `['mesos']['slave']['flags']` hashes.

The valid list of Mesos options may be found at:
https://github.com/apache/mesos/blob/master/docs/configuration.md

## Recipes

### default
The default mesos recipe will run mesos::install.

### install
The install recipe installs the specified version of the mesosphere mesos
RPM or Debian package and installs it.  It’s also configured to stop both
mesos-master and mesos-slave init files so that they don't automatically
start on server restart.

### master
The master recipe runs mesos::install as well as creating several
mesos-master configuration files that are used at startup.  This recipe also
uses the zookeeper attributes and/or exhibitor attributes to configure the
mesos-master using zookeeper.  Lastly it sets the mesos-master init config to
'start' so that mesos-master is started on server restart.

### slave
The slave recipe runs mesos::install as well as creating several
mesos-slave configuration files that are used at startup.  This recipe also
uses the zookeeper attributes and/or exhibitor attributes to configure the
mesos-slave using zookeeper.  Lastly it sets the mesos-slave init config to
'start' so that mesos-slave is started on server restart.

### repo
The repo recipe contains logic for setting up Mesosphere debian and RPM
repositories.

Dependencies
------------

The following cookbooks are dependencies:

* [apt][]
* [yum][]
* [java][]

The following cookbooks are suggested:

* [exhibitor][] (used for discovering ZooKeeper ensembles via [Netflix Exhibitor][])

Usage
-----

Here is a sample role for configuring a Mesos master in a ZooKeeper backed
production mode.

```YAML
chef_type:           role
default_attributes:
description:
env_run_lists:
json_class:          Chef::Role
name:                mesos_master
override_attributes:
  mesos:
    version: 1.0.1
    master:
      flags:
        cluster: 'mesos-sandbox'
        zk: 'zk://127.0.0.1:2181/mesos'
run_list:
  recipe[mesos::master]
```

Here is a sample role for creating a Mesos slave node with a seperate ZooKeeper
ensemble dynamically discovered via Netflix Exhibitor:
```YAML
chef_type:           role
default_attributes:
description:
env_run_lists:
json_class:          Chef::Role
name:                mesos_slave
override_attributes:
  mesos:
    version: 1.0.1
    slave:
      flags:
        master: 'zk://127.0.0.1:2181/mesos'
run_list:
  recipe[mesos::slave]
```

Here is a role for creating a Mesos master node with Marathon and Chronos
```JSON
{
  "name": "mesos-master",
  "environment": "test_env",
  "description": "Role for mesos master servers",
  "json_class": "Chef::Role",
  "default_attributes": {
    "mesos": {
      "version": "1.3.0-2.0.3",
      "master": {
        "flags": {
          "cluster": "exampleStageCluster",
          "zk": "zk://mesos-master1-test.example.org:2181,mesos-master2-test.example.org:2181,mesos-master3-test.example.org:2181/mesos",
          "quorum": "2",
          "work_dir": "/var/lib/mesos"
        }
      },
      "marathon": {
        "install": "true",
        "flags": {
          "zk": "zk://mesos-master1-test.example.org:2181,mesos-master2-test.example.org:2181,mesos-master3-test.example.org:2181/marathon",
          "master": "zk://mesos-master1-test.example.org:2181,mesos-master2-test.example.org:2181,mesos-master3-test.example.org:2181/mesos"
        }
      },
      "chronos": {
        "install": "true",
        "flags": {
          "http_port": "4400",
          "zk_hosts": "zk://mesos-master1-test.example.org:2181,mesos-master2-test.example.org:2181,mesos-master3-test.example.org:2181",
          "master": "zk://mesos-master1-test.example.org:2181,mesos-master2-test.example.org:2181,mesos-master3-test.example.org:2181/mesos"
        }
      }
    }
  },
  "override_attributes": {

  },
  "chef_type": "role",
  "run_list": [
    "mesos::master"
  ],
  "env_run_lists": {

  }
}
```

Development
-----------
Please see the [Contributing](CONTRIBUTING.md) and [Issue Reporting](ISSUES.md) Guidelines.

License and Author
------------------
* Author: [Ray Rodriguez](https://github.com/rayrod2030)(rayrod2030@gmail.com)
* Author: [Robert Veznaver](https://github.com/rveznaver)(robert.veznaver@gmail.com)

Copyright 2015 Medidata Solutions Worldwide

Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

[Apache Mesos]: http://mesos.apache.org
[Netflix Exhibitor]: https://github.com/Netflix/exhibitor
[Mesosphere]: http://mesosphere.io
[Medidata Solutions]: http://www.mdsol.com
[exhibitor]: https://github.com/SimpleFinance/chef-exhibitor
[apt]: https://github.com/opscode-cookbooks/apt
[yum]: https://github.com/chef-cookbooks/yum
[java]: https://github.com/agileorbit-cookbooks/java


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [3.6.1] - 2017-08-04
### Added
- Marathon installation

## [3.6.0] - 2016-12-05
### Fixed
- Mesos 1.0.0 and 1.0.1 require the --work_dir flag to be specified

### Changed
- Default to Mesos 1.1.0
- Default Java version to 8
- Unlock yum cookbook version
- Platform list updated to match Mesosphere packages

## [3.5.1] - 2015-12-23
### Fixed
- Fix Amazon Linux support

### Changed
- Set up yum repo with `yum_repository` provider
- Remove dependency on chef-sugar

## [3.5.0] - 2015-12-21
### Added
- Systemd support
- Dynamic configuration guard
- Set default `cgroups_hierarchy` attribute

### Changed
- Improved upstart script to start after network (@kamaradclimber)
- Default to Mesos 0.25.0
- Removed support for Mesos < 0.23.0
- Refactored testing suite
- Platform list updated to match Mesosphere package
- Tested with Chef 12.5.1

## [3.4.0] - 2015-08-25
### Added
- Added support for Apache Mesos 0.23.0.

### Changed
- Removed support for Ubuntu 13.10.
- Removed support for Mesos 0.20.x.
- Changed all compile_time calls to at_compile_time for Chef 12.2.x compatibility with ChefSugar.

## [3.3.0] - 2015-07-05
### Added
- Added support for Apache Mesos 0.22.1.
- Added vagrant-wrapper in gemfile for testing.
- Set ulimit for init service if defined.
- Install mesosphere packages from a package mirror.

### Changed
- Switched to using port 80 for Ubuntu Keyservers to get through corporate firewalls.

## [3.2.3] - 2015-03-29
### Fixed
- Fixed master and slave recipes to use node override for zk endpoints.

## [3.2.2] - 2015-03-29
### Fixed
- Fixed master and slave recipes to use node instead of default to set zk endpoints.

## [3.2.1] - 2015-03-29
### Fixed
- Fixed zk exhibitor discovery logic to use new master and slave flags.

## [3.2.0] - 2015-03-29
### Changed
- Updated default Mesos version to 0.21.1.
- Added support for Mesos 0.22.0.
- Removed support for Mesos 0.19.x.

## [3.1.0] - 2015-03-27
### Changed
- Refactor greatly simplifies Mesos wrapper script.
- Configuration now passed primarily via command line flags.
- Switched to Upstart & Sysvinit for Mesos init.
- Metadata uses suggests instead of recommends for the exhibitor cookbook.

### Contributors for this release:

- [Robert Vežnaver](https://github.com/rveznaver) - Amazing Mesos wrapper refactor.

Thank You!

## [3.0.0] - 2015-03-11
### Changed
- Mesos is now installed via the Mesosphere RPM and deb package repositories.
- Added validation for Mesos flags to ensure that Mesos starts with configured
flags.
- Added support for Mesos 0.21.1.
- Added all Mesos attributes which can be configured.
- Added libsvn1 depedency for newer Mesos versions.
- Updated rubocop style file.

### Fixed
- Fixed a bug with the ZK exhibitor discovery mechanism.
- Fixed redhat family name in install recipe.

### Removed
- Removed experimental docker recipe.
- Removed EC2 hostname logic.
- Removed support for Mesos 0.19.0.

## [2.0.0] - 2014-11-14
### Changed
- Writes all attributes under `node['mesos']['common]`/`node['mesos']['master']` to respective config files
- Adds support for Ubuntu 14.04 (Trusty Tahr)
- Adds support for Debian 7 (Wheezy)

## [1.0.6] - 2014-02-19
### Changed
- Re-added default-jre-headless as required package due to breaking issues with
  mesos debian package when used with a non-openjdk java like Oracle.
- Added some serverspecs to verify working cookbook with oracle jdk 7.

## [1.0.5] - 2014-02-15
### Changed

- Removed default-jre-headless as required package
- Added slave checkpointing defaults
- Added kitchen-ec2 configuration for CI

## [1.0.4] - 2014-02-07
### Changed

- Add support for RHEL linux family
- Updated integration tests (serverspec)

## [1.0.3] - 2014-02-03
### Changed

- Fix zk_string generating logic to strip triailing comma
- Add exhibitor discover zookeepers retry logic

## [1.0.2] - 2014-01-31
### Changed

- Fix docker-mesos egg url

## [1.0.1] - 2014-01-13
### Changed

- Update default mesos version to 0.15.0
- Use attribute for mesos python bindings egg

## [1.0.0] - 2013-12-12
### Changed

- Public release of cookbook
- Added experimental docker-mesos recipe
- Updated mesos default version to 0.14.2

## [0.1.0] - 2013-10-08
### Changed

- Initial Cookbook release

[3.6.0]: https://github.com/mdsol/mesos_cookbook/compare/3.5.1...3.6.0
[3.5.1]: https://github.com/mdsol/mesos_cookbook/compare/3.5.0...3.5.1
[3.5.0]: https://github.com/mdsol/mesos_cookbook/compare/3.4.0...3.5.0
[3.4.0]: https://github.com/mdsol/mesos_cookbook/compare/3.3.0...3.4.0
[3.3.0]: https://github.com/mdsol/mesos_cookbook/compare/3.2.3...3.3.0
[3.2.3]: https://github.com/mdsol/mesos_cookbook/compare/3.2.2...3.2.3
[3.2.2]: https://github.com/mdsol/mesos_cookbook/compare/3.2.1...3.2.2
[3.2.1]: https://github.com/mdsol/mesos_cookbook/compare/3.2.0...3.2.1
[3.2.0]: https://github.com/mdsol/mesos_cookbook/compare/3.1.0...3.2.0
[3.1.0]: https://github.com/mdsol/mesos_cookbook/compare/3.0.0...3.1.0
[3.0.0]: https://github.com/mdsol/mesos_cookbook/compare/2.0.0...3.0.0
[2.0.0]: https://github.com/mdsol/mesos_cookbook/compare/1.0.6...2.0.0
[1.0.6]: https://github.com/mdsol/mesos_cookbook/compare/1.0.5...1.0.6
[1.0.5]: https://github.com/mdsol/mesos_cookbook/compare/1.0.4...1.0.5
[1.0.4]: https://github.com/mdsol/mesos_cookbook/compare/1.0.3...1.0.4
[1.0.3]: https://github.com/mdsol/mesos_cookbook/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/mdsol/mesos_cookbook/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/mdsol/mesos_cookbook/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/mdsol/mesos_cookbook/compare/0.1.0...1.0.0
[0.1.0]: https://github.com/mdsol/mesos_cookbook/compare/0.1.0...0.1.0


recipes/master.rb
#
# Cookbook Name:: mesos
# Recipe:: master
#
# Copyright (C) 2015 Medidata Solutions, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

class Chef::Recipe
  include MesosHelper
end

include_recipe 'mesos::install'

# Mesos configuration validation
ruby_block 'mesos-master-configuration-validation' do
  block do
    # Get Mesos --help
    help = Mixlib::ShellOut.new("#{node['mesos']['master']['bin']} --help")
    help.run_command
    help.error!
    # Extract options
    options = help.stdout.strip.scan(/^  --(?:\[no-\])?(\w+)/).flatten - ['help']
    # Check flags are in the list
    node['mesos']['master']['flags'].keys.each do |flag|
      unless options.include?(flag)
        Chef::Application.fatal!("Invalid Mesos configuration option: #{flag}. Aborting!", 1000)
      end
    end
  end
end

# ZooKeeper Exhibitor discovery
if node['mesos']['zookeeper_exhibitor_discovery'] && node['mesos']['zookeeper_exhibitor_url']
  zk_nodes = MesosHelper.discover_zookeepers_with_retry(node['mesos']['zookeeper_exhibitor_url'])

  if zk_nodes.nil?
    Chef::Application.fatal!('Failed to discover zookeepers. Cannot continue.')
  end

  node.override['mesos']['master']['flags']['zk'] = 'zk://' + zk_nodes['servers'].sort.map { |s| "#{s}:#{zk_nodes['port']}" }.join(',') + '/' + node['mesos']['zookeeper_path']
end

# Mesos master configuration wrapper
template 'mesos-master-wrapper' do
  path '/etc/mesos-chef/mesos-master'
  owner 'root'
  group 'root'
  mode '0750'
  source 'wrapper.erb'
  variables(env:    node['mesos']['master']['env'],
            bin:    node['mesos']['master']['bin'],
            flags:  node['mesos']['master']['flags'],
            syslog: node['mesos']['master']['syslog'])
end

# Mesos master service definition
service 'mesos-master' do
  case node['mesos']['init']
  when 'systemd'
    provider Chef::Provider::Service::Systemd
  when 'sysvinit_debian'
    provider Chef::Provider::Service::Init::Debian
  when 'upstart'
    provider Chef::Provider::Service::Upstart
  end
  supports status: true, restart: true
  subscribes :restart, 'template[mesos-master-init]'
  subscribes :restart, 'template[mesos-master-wrapper]'
  action [:enable, :start]
end

if node['mesos']['marathon']['install']
  include_recipe 'mesos::marathon'
end

if node['mesos']['chronos']['install']
  include_recipe 'mesos::chronos'
end


attributes/default.rb
# Default Java version
default['java']['jdk_version'] = '8'

# Use Mesosphere repo
default['mesos']['repo']       = true

# Mesosphere Mesos version.
default['mesos']['version']    = '1.1.0'

# Init system to use
default['mesos']['init']       = case node['platform']
                                 when 'debian'
                                   node['platform_version'].to_i >= 8 ? 'systemd' : 'sysvinit_debian'
                                 when 'ubuntu'
                                   node['platform_version'].to_f >= 15.04 ? 'systemd' : 'upstart'
                                 when 'redhat', 'centos', 'scientific', 'oracle' # ~FC024
                                   node['platform_version'].to_i >= 7 ? 'systemd' : 'upstart'
                                 else 'upstart'
                                 end

#
# Mesos MASTER configuration
#

# Mesos master binary location.
default['mesos']['master']['bin']                   = '/usr/sbin/mesos-master'

# Environmental variables set before calling the mesos master process.
default['mesos']['master']['env']['ULIMIT']         = '-n 16384'

# Send stdout and stderr to syslog.
default['mesos']['master']['syslog']                = true

# Mesos master command line flags.
# http://mesos.apache.org/documentation/latest/configuration/
default['mesos']['master']['flags']['port']          = 5050
default['mesos']['master']['flags']['log_dir']       = '/var/log/mesos'
default['mesos']['master']['flags']['logging_level'] = 'INFO'
default['mesos']['master']['flags']['cluster']       = 'MyMesosCluster'
default['mesos']['master']['flags']['work_dir']      = '/tmp/mesos'

#
# Mesos SLAVE configuration
#

# Mesos slave binary location.
default['mesos']['slave']['bin']                    = '/usr/sbin/mesos-slave'

# Environmental variables set before calling the mesos-slave process.
default['mesos']['slave']['env']['ULIMIT']          = '-n 16384'

# Send stdout and stderr to syslog.
default['mesos']['slave']['syslog']                 = true

# Mesos slave command line flags
# http://mesos.apache.org/documentation/latest/configuration/
default['mesos']['slave']['flags']['port']          = 5051
default['mesos']['slave']['flags']['log_dir']       = '/var/log/mesos'
default['mesos']['slave']['flags']['logging_level'] = 'INFO'
default['mesos']['slave']['flags']['work_dir']      = '/tmp/mesos'
default['mesos']['slave']['flags']['isolation']     = 'posix/cpu,posix/mem'
default['mesos']['slave']['flags']['master']        = 'localhost:5050'
default['mesos']['slave']['flags']['strict']        = true
default['mesos']['slave']['flags']['recover']       = 'reconnect'

# Workaround for setting default cgroups hierarchy root
default['mesos']['slave']['flags']['cgroups_hierarchy'] = if node['mesos']['init'] == 'systemd'
                                                            '/sys/fs/cgroup'
                                                          else
                                                            '/cgroup'
                                                          end

# Use the following options if you are using Exhibitor to manage Zookeeper
# in your environment.

# Zookeeper path that Mesos will use to write to.
default['mesos']['zookeeper_path']                      = 'mesos'

# Flag to enable Zookeeper ensemble discovery via Netflix Exhibitor.
default['mesos']['zookeeper_exhibitor_discovery']       = false

# Netflix Exhibitor ZooKeeper ensemble url.
default['mesos']['zookeeper_exhibitor_url']             = nil

#
# Marathon framework
#
default['mesos']['marathon']['install']         = false
default['mesos']['marathon']['version']         = '1.4.5-1.0.654.ubuntu1604'
default['mesos']['marathon']['flags']['master'] = 'zk://localhost:2181/marathon'
default['mesos']['marathon']['flags']['zk']     = 'zk://localhost:2181/mesos'


#
# Chronos framework
#
default['mesos']['chronos']['install']            = false
default['mesos']['chronos']['version']            = '2.5.0-0.1.20170628182950.ubuntu1604'
default['mesos']['chronos']['flags']['http_port'] = 4400
default['mesos']['chronos']['flags']['zk_hosts']  = 'zk://localhost:2181'
default['mesos']['chronos']['flags']['master']    = 'zk://localhost:2181/mesos'
